sdk/communication/Azure.Communication.Calling.Server/src/PartitionedDownloader.cs

```diff
@@ -0,0 +1,323 @@
+                // small blob, we'll get the whole thing in one shot.  If it's
+                // a large blob, we'll get its full size in Content-Range and
+                // can keep downloading it in segments.
+                var initialRange = new HttpRange(0, _initialRangeSize);
+                Response<Stream> initialResponse;
+
+                try
+                {
+                    initialResponse = _client.DownloadStreaming(
+                        endpoint,
+                        initialRange,
+                        cancellationToken);
+                }
+                catch (RequestFailedException ex) when (ex.ErrorCode == "Invalid Range")
+                {
+                    initialResponse = _client.DownloadStreaming(
+                        endpoint,
+                        range: default,
+                        cancellationToken);
+                }
+
+                // Copy the first segment to the destination stream
+                CopyTo(initialResponse, destination, cancellationToken);
+
+                // If the first segment was the entire blob, we're finished now
+                long initialLength = ParseResponseContentLength(initialResponse);
+                long totalLength = ParseRangeTotalLength(initialResponse);
+                if (initialLength == totalLength)
+                {
+                    return initialResponse.GetRawResponse();
+                }
+
+                // Download each of the remaining ranges in the blob
+                foreach (HttpRange httpRange in GetRanges(initialLength, totalLength))
+                {
+                    // Don't need to worry about 304s here because the ETag
+                    // condition will turn into a 412 and throw a proper
+                    // RequestFailedException
+                    Response<Stream> result = _client.DownloadStreaming(
+                        endpoint,
+                        httpRange,
+                        cancellationToken);
+                    CopyTo(result.Value, destination, cancellationToken);
+                }
+
+                return initialResponse.GetRawResponse();
+            }
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
+            finally
+            {
+                scope.Dispose();
+            }
+        }
+
+        private static long ParseRangeTotalLength(Response<Stream> response)
+        {
+            response.GetRawResponse().Headers.TryGetValue("Content-Range", out string range);
+
+            if (range == null)
+            {
+                return 0;
+            }
+            int lengthSeparator = range.IndexOf("/", StringComparison.InvariantCultureIgnoreCase);
+            if (lengthSeparator == -1)
+            {
+                throw new ArgumentException("Could not obtain the total length from HTTP range " + range);
```

Comment:
`ArgumentException` does not sound like the right type to throw as `range` comes from the response.